desc/singularities.py

```diff
@@ -0,0 +1,733 @@
+    dz = abs(zeta - zeta0)
+    dt = jnp.minimum(dt, 2 * np.pi - dt)
+    dz = jnp.minimum(dz, 2 * np.pi - dz)
+    return 2 / s * jnp.sqrt((dt / dtheta) ** 2 + (dz / dzeta) ** 2)
+
+
+def _nonsingular_part(eval_data, eval_grid, src_data, src_grid, s, kernel, loop=False):
+    """Integrate kernel over non-singular points."""
+    assert isinstance(src_grid.NFP, int)
+    src_theta = src_grid.nodes[:, 1]
+    src_zeta = src_grid.nodes[:, 2]
+    src_dtheta = src_grid.spacing[:, 1]
+    src_dzeta = src_grid.spacing[:, 2] / src_grid.NFP
+    eval_theta = jnp.asarray(eval_grid.nodes[:, 1])
+    eval_zeta = jnp.asarray(eval_grid.nodes[:, 2])
+    w = src_grid.weights * src_data["|e_theta x e_zeta|"] / src_grid.NFP
+    h_t = jnp.mean(src_dtheta)
+    h_z = jnp.mean(src_dzeta)
+
+    keys = kernel.keys
+
+    def nfp_loop(j, f_data):
+        # calculate effects at all eval pts from all src pts on a single field period,
+        # summing over field periods
+        f, src_data = f_data
+        src_data["zeta"] = (src_zeta + j * 2 * np.pi / src_grid.NFP) % (2 * np.pi)
+
+        # nest this def to avoid having to pass the modified src_data around the loop
+        # easier to just close over it and let JAX figure it out
+        def eval_pt_vmap(i):
+            # this calculates the effect at a single evaluation point, from all others
+            # in a single field period. vmap this to get all pts
+            k = kernel(
+                {key: val[i] for key, val in eval_data.items() if key in keys},
+                src_data,
+            ).reshape((1, src_grid.num_nodes, kernel.ndim))
+
+            rho = _rho(
+                src_theta,
+                src_data["zeta"],  # to account for different field periods
+                eval_theta[i],
+                eval_zeta[i],
+                h_t,
+                h_z,
+                s,
+            )
+
+            eta = _chi(rho)
+            k = (1 - eta)[None, :, None] * k
+            f_temp = jnp.sum(k * w[None, :, None], axis=1)
+            return f_temp
+
+        def eval_pt_loop(i, fj):
+            # this calculates the effect at a single evaluation point, from all others
+            # in a single field period. loop this to get all pts
+            f_temp = eval_pt_vmap(i)
+            return put(fj, i, f_temp.reshape(fj[i].shape))
+
+        # vmap for inner part found more efficient than fori_loop, especially on gpu,
+        # but for jacobian looped seems to be better and less memory
+        if loop:
+            fj = fori_loop(0, eval_grid.num_nodes, eval_pt_loop, jnp.zeros_like(f))
+        else:
+            fj = vmap(eval_pt_vmap)(jnp.arange(eval_grid.num_nodes))
+
+        f += fj.reshape((eval_grid.num_nodes, kernel.ndim))
+        return f, src_data
+
+    f = jnp.zeros((eval_grid.num_nodes, kernel.ndim))
+    f, _ = fori_loop(0, src_grid.NFP, nfp_loop, (f, src_data))
+
+    # we sum distance vectors, so they need to be in xyz for that to work
```

Comment:
where is this sum happening? in the kernel evaluations?

Reply:
It happens when we sum the contribution from different source points